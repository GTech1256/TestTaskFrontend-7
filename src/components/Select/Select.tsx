import React, { ChangeEvent, useCallback } from 'react'


const DEFAULT_VALUE = 'DEFAULT'

type OptionType = {
  value: string,
  label: string,
  isSelected?: boolean
}

export type Props = {
  className?: string
  title?: string
  name: string
  isLoading?: boolean
  value?: OptionType['value']
  options: OptionType[]
  onChange: (newValue: OptionType['value'] | undefined) => void
}

const renderOption = ({
  label,
  value,
}: OptionType) => <option key={label} value={value}>{label}</option>

const Select = ({
  className = '',
  title,
  name,
  options,
  onChange,
  isLoading = false,
}: Props) => {
  // Возвращает undefined, если выбрано дефолтное поле
  const handleChange = useCallback<(event: ChangeEvent<HTMLSelectElement>) => void>(({ target: { value } }) => {
    onChange(value === DEFAULT_VALUE ? undefined : value)
  }, [onChange])

  return (
    <div className={`input-group mb-3 ${className}`}>
      {title && (
        <div className="input-group-prepend">
          <label className="input-group-text" htmlFor={name}>{title}</label>
        </div>
      )}
      <select
        className="custom-select"
        id={name}
        onChange={handleChange}
        defaultValue={DEFAULT_VALUE}
        disabled={isLoading}
      >
        <option value={DEFAULT_VALUE}>Фильтр по лицензии...</option>
        {options.map(renderOption)}
      </select>
    </div>
  )
}

export default Select;