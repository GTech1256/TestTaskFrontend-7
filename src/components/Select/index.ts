export type { Props as SelectProps } from './Select'
export { default } from './Select'