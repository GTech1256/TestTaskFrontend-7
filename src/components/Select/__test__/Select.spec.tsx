import React from 'react';
import renderer from 'react-test-renderer';

import Select, { SelectProps } from '../index'


const props: SelectProps = {
  className: 'test-class-name',
  name: 'testSelect',
  title: 'testTitle',
  options: [
    { value: 'value1', label: 'lebel1', isSelected: true },
    { value: 'value3', label: 'lebel2' }
  ],
  onChange: () => {}
}

describe('<Select />', () => {
  it('should render correct', () => {
    const tree = renderer
      .create(<Select {...props} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  })
})