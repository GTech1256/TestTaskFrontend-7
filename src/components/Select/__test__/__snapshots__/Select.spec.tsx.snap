// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render correct 1`] = `
<div
  className="input-group mb-3 test-class-name"
>
  <div
    className="input-group-prepend"
  >
    <label
      className="input-group-text"
      htmlFor="testSelect"
    >
      testTitle
    </label>
  </div>
  <select
    className="custom-select"
    defaultValue="DEFAULT"
    disabled={false}
    id="testSelect"
    onChange={[Function]}
  >
    <option
      value="DEFAULT"
    >
      Фильтр по лицензии...
    </option>
    <option
      value="value1"
    >
      lebel1
    </option>
    <option
      value="value3"
    >
      lebel2
    </option>
  </select>
</div>
`;
